src/server/snap.rs

```diff
@@ -150,23 +155,36 @@ fn send_snap(
         }
     };
 
+    fail_point!("snapshot_before_send", |_| Err(box_err!(
+        "abort snapshot by fp"
+    )));
+
     let cb = ChannelBuilder::new(env)
         .stream_initial_window_size(cfg.grpc_stream_initial_window_size.0 as usize)
         .keepalive_time(cfg.grpc_keepalive_time.0)
         .keepalive_timeout(cfg.grpc_keepalive_timeout.0)
         .default_compression_algorithm(cfg.grpc_compression_algorithm());
 
-    let channel = security_mgr.connect(cb, addr);
-    let client = TikvClient::new(channel);
-    let (sink, receiver) = client.snapshot()?;
-
-    let send = chunks.forward(sink).map_err(Error::from);
-    let send = send.and_then(|(s, _)| receiver.map_err(Error::from).map(|_| s))
-        .then(move |result| {
+    Ok(addr.map_err(|e| Error::Other(box_err!(e)))
+        .and_then(move |addr| {
+            let channel = security_mgr.connect(cb, &addr);
+            let client = TikvClient::new(channel);
+            client
+                .snapshot()
+                .map(|(l, r)| (client, l, r))
+                .map_err(Error::from)
+        })
+        .and_then(|(client, sink, receiver)| {
+            chunks
+                .forward(sink)
+                .map(|(s, _)| ((client, s), receiver))
+                .map_err(Error::from)
+        })
+        .and_then(|(res, receiver)| receiver.map(|_| res).map_err(Error::from))
+        .then(move |res| {
             send_timer.observe_duration();
             drop(deregister);
-            drop(client);
-            result.map(|s| {
+            res.map(|(_, s)| {
```

Comment:
Don't need to keep `client` undroped until here?

Reply:
The content of res is `(client, chunks)`.